### PR TITLE
Made tweaks to core architecture.

### PR DESCRIPTION
**Pull request by patriksvensson**
I've divided the core into different, more distinguished parts that has clearer areas of responsibility.
This will make it easier with things like adopting the new Roslyn bits.

`IScriptHost`
  - Contains the environment visible to scripts.

`ICakeContext`
  - Acts as a context for scripts aliases and extensions.

`ICakeEngine`
  - Responsible for execution planning, workflow and rules.

`IExecutionStrategy`
  - Handles task execution.

### CODE REVIEW

File: src/Cake.Core/CakeContext.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        /// </summary>
+        /// <param name="fileSystem">The file system.</param>
+        /// <param name="environment">The environment.</param>
+        /// <param name="globber">The globber.</param>
+        /// <param name="log">The log.</param>
+        /// <param name="arguments">The arguments.</param>
+        /// <param name="processRunner">The process runner.</param>
+        /// <param name="toolResolvers">The tool resolvers.</param>
+        public CakeContext(
+            IFileSystem fileSystem,
```
Comment (devlead) [inline src/Cake.Core/CakeContext.cs]: Super minor `filesystem`, `environment`, `globber`, `log` & `arguments` have trailing spaces.

Comment (patriksvensson) [inline src/Cake.Core/CakeContext.cs]: Will fix :smile:

File: src/Cake.Core/CakeEngine.cs [hunk 2]
```diff
@@ -14,34 +13,10 @@ namespace Cake.Core
...
-
-        /// <summary>
-        /// Gets the environment.
-        /// </summary>
-        /// <value>The environment.</value>
-        public ICakeEnvironment Environment
-        {
-            get { return _environment; }
-        }
+        private Action _teardownAction;
```
Comment (devlead) [inline src/Cake.Core/CakeEngine.cs]: trailing spaces after `_teardownAction;`

File: src/Cake.Core/DefaultExecutionStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        /// <summary>
+        /// Invokes the finally handler.
+        /// </summary>
+        /// <param name="action">The action.</param>
+        public void InvokeFinally(Action action)
+        {
+            if (action != null)
+            {
+                action();
```
Comment (devlead) [inline src/Cake.Core/DefaultExecutionStrategy.cs]: trailing spaces after `action();` and the next row `}`

File: src/Cake.Tests/Unit/Scripting/BuildScriptHostTests.cs [hunk 4]
```diff
@@ -32,9 +33,10 @@ public void Should_Print_Report()
                 var report = new CakeReport();
                 report.Add("Target", TimeSpan.FromSeconds(1));
                 var engine = Substitute.For<ICakeEngine>();
-                engine.RunTarget("Target").Returns(report);
+                var context = Substitute.For<ICakeContext>();
+                engine.RunTarget(context, "Target").Returns(report);
```
Comment (devlead) [inline src/Cake.Tests/Unit/Scripting/BuildScriptHostTests.cs]: trailing spaces after `Returns(report);`

File: src/Cake/Scripting/BuildScriptHost.cs [hunk 5]
```diff
@@ -14,9 +14,10 @@ public sealed class BuildScriptHost : ScriptHost
         /// Initializes a new instance of the <see cref="BuildScriptHost"/> class.
         /// </summary>
         /// <param name="engine">The engine.</param>
+        /// <param name="context">The context.</param>
         /// <param name="reportPrinter">The report printer.</param>
-        public BuildScriptHost(ICakeEngine engine, ICakeReportPrinter reportPrinter) 
-            : base(engine)
+        public BuildScriptHost(ICakeEngine engine, ICakeContext context, ICakeReportPrinter reportPrinter)
```
Comment (devlead) [inline src/Cake/Scripting/BuildScriptHost.cs]: trailing space

**Comment by gep13** [comment]
Cough...

https://visualstudiogallery.msdn.microsoft.com/a204e29b-1778-4dae-affd-209bea658a59

**Comment by devlead** [comment]
@patriksvensson only had some super minor formatting iisues, really like this PR, especially cleanup of `CakeEngine` and the introduction of  `CakeContext`, makes allot of sense.

**Comment by devlead** [comment]
@gep13 good one :+1:  could spot a mile away just by having visible white space in my editor (as all should, along with 4 spaces instead of tabs;))

**Comment by patriksvensson** [comment]
@gep13 Haha, perfect! Will install that one! :smile:

**Comment by gep13** [comment]
Yip, it is one of my new favourites :smile_cat: makes all trailing whitespace really obvious, so no excuses to leave them in.

File: src/Cake.Core.Tests/Fixtures/ScriptHostFixture.cs [hunk 6]
```diff
@@ -35,18 +36,20 @@ public ScriptHostFixture()
             Globber = Substitute.For<IGlobber>();
             Arguments = Substitute.For<ICakeArguments>();
 
+            Context = Substitute.For<ICakeContext>();
+            Context.Arguments.Returns(Arguments);
+            Context.Environment.Returns(Environment);
+            Context.FileSystem.Returns(FileSystem);
+            Context.Globber.Returns(Globber);
+            Context.Log.Returns(Log);
```
Comment (devlead) [inline src/Cake.Core.Tests/Fixtures/ScriptHostFixture.cs]: Trailing spaces after `(Log);`

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 7]
```diff
@@ -227,10 +200,12 @@ public void Should_Not_Skip_Tasks_Where_Criterias_Are_Fulfilled()
             public void Should_Throw_If_Target_Was_Not_Found()
             {
                 // Given
-                var engine = new CakeEngineFixture().CreateEngine();
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("Run-Some-Tests"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 8]
```diff
@@ -241,11 +216,13 @@ public void Should_Throw_If_Target_Was_Not_Found()
...
                 // Given
-                var engine = new CakeEngineFixture().CreateEngine();
-                engine.Task("A").Does(() => { throw new InvalidOperationException("Whoopsie"); });
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
+                engine.RegisterTask("A").Does(() => { throw new InvalidOperationException("Whoopsie"); });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 9]
```diff
@@ -268,13 +246,15 @@ public void Should_Invoke_Task_Error_Handler_If_Exception_Is_Thrown()
...
-                engine.Task("A")
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
+                engine.RegisterTask("A")
                     .Does(() => { throw new InvalidOperationException("Whoopsie"); })
                     .OnError(exception => { invoked = true; });
 
                 // When
-                Record.Exception(() => engine.RunTarget("A"));
+                Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 10]
```diff
@@ -284,13 +264,15 @@ public void Should_Invoke_Task_Error_Handler_If_Exception_Is_Thrown()
...
-                engine.Task("A")
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
+                engine.RegisterTask("A")
                     .Does(() => { throw new InvalidOperationException("Whoopsie"); })
                     .OnError(exception => { throw new InvalidOperationException("Totally my fault"); });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 11]
```diff
@@ -303,12 +285,13 @@ public void Should_Log_Exception_Handled_By_Error_Handler()
...
                 var fixture = new CakeEngineFixture();
                 var engine = fixture.CreateEngine();
-                engine.Task("A")
+                engine.RegisterTask("A")
                     .Does(() => { throw new InvalidOperationException("Whoopsie"); })
                     .OnError(exception => { throw new InvalidOperationException("Totally my fault"); });
 
                 // When
-                Record.Exception(() => engine.RunTarget("A"));
+                Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 12]
```diff
@@ -392,12 +379,13 @@ public void Should_Run_Teardown_After_Last_Running_Task_Even_If_Task_Failed()
...
-                engine.Setup(() => { });
-                engine.Teardown(() => { });
-                engine.Task("A").Does(() => { throw new InvalidOperationException("Fail"); });
+                engine.RegisterSetupAction(() => { });
+                engine.RegisterTeardownAction(() => { });
+                engine.RegisterTask("A").Does(() => { throw new InvalidOperationException("Fail"); });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 13]
```diff
@@ -413,12 +401,13 @@ public void Should_Run_Teardown_If_Setup_Failed()
...
-                engine.Setup(() => { throw new InvalidOperationException("Fail"); });
-                engine.Teardown(() => { });
-                engine.Task("A").Does(() => { });
+                engine.RegisterSetupAction(() => { throw new InvalidOperationException("Fail"); });
+                engine.RegisterTeardownAction(() => { });
+                engine.RegisterTask("A").Does(() => { });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 14]
```diff
@@ -434,12 +423,13 @@ public void Should_Throw_Exception_Thrown_From_Setup_Action_If_Both_Setup_And_Te
...
-                engine.Setup(() => { throw new InvalidOperationException("Setup"); });
-                engine.Teardown(() => { throw new InvalidOperationException("Teardown"); });
-                engine.Task("A").Does(() => { });
+                engine.RegisterSetupAction(() => { throw new InvalidOperationException("Setup"); });
+                engine.RegisterTeardownAction(() => { throw new InvalidOperationException("Teardown"); });
+                engine.RegisterTask("A").Does(() => { });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 15]
```diff
@@ -455,11 +445,12 @@ public void Should_Throw_Exception_Occuring_In_Teardown_If_No_Previous_Exception
...
                 var expected = new InvalidOperationException("Teardown");
 
-                engine.Teardown(() => { throw expected; });
-                engine.Task("A").Does(() => { });
+                engine.RegisterTeardownAction(() => { throw expected; });
+                engine.RegisterTask("A").Does(() => { });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 16]
```diff
@@ -472,12 +463,13 @@ public void Should_Log_Teardown_Exception_If_Both_Setup_And_Teardown_Actions_Thr
...
-                engine.Setup(() => { throw new InvalidOperationException("Setup"); });
-                engine.Teardown(() => { throw new InvalidOperationException("Teardown"); });
-                engine.Task("A").Does(() => { });
+                engine.RegisterSetupAction(() => { throw new InvalidOperationException("Setup"); });
+                engine.RegisterTeardownAction(() => { throw new InvalidOperationException("Teardown"); });
+                engine.RegisterTask("A").Does(() => { });
 
                 // When
-                Record.Exception(() => engine.RunTarget("A"));
+                Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 17]
```diff
@@ -490,11 +482,12 @@ public void Should_Exception_Thrown_From_Task_If_Both_Task_And_Teardown_Actions_
...
                 var engine = fixture.CreateEngine();
 
-                engine.Teardown(() => { throw new InvalidOperationException("Teardown"); });
-                engine.Task("A").Does(() => { throw new InvalidOperationException("Task"); });
+                engine.RegisterTeardownAction(() => { throw new InvalidOperationException("Teardown"); });
+                engine.RegisterTask("A").Does(() => { throw new InvalidOperationException("Task"); });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 18]
```diff
@@ -509,11 +502,12 @@ public void Should_Log_Teardown_Exception_If_Both_Task_And_Teardown_Actions_Thro
...
                 var engine = fixture.CreateEngine();
 
-                engine.Teardown(() => { throw new InvalidOperationException("Teardown"); });
-                engine.Task("A").Does(() => { throw new InvalidOperationException("Task"); });
+                engine.RegisterTeardownAction(() => { throw new InvalidOperationException("Teardown"); });
+                engine.RegisterTask("A").Does(() => { throw new InvalidOperationException("Task"); });
 
                 // When
-                Record.Exception(() => engine.RunTarget("A"));
+                Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 19]
```diff
@@ -595,13 +594,15 @@ public void Should_Execute_Error_Reporter_Before_Error_Handler_If_Task_Succeeds(
...
-                engine.Task("A")
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
+                engine.RegisterTask("A")
                     .Does(() => { throw new InvalidOperationException("Task"); })
                     .ReportError(ex => { throw new InvalidOperationException("Report"); });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 20]
```diff
@@ -611,14 +612,16 @@ public void Should_Swallow_Exceptions_Thrown_In_Error_Reporter()
...
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
+                engine.RegisterTask("A")
                     .Does(() => { throw new InvalidOperationException("Task"); })
                     .OnError(() => { throw new InvalidOperationException("Error"); })
                     .ReportError(ex => { throw new InvalidOperationException("Report"); });
 
                 // When
-                var result = Record.Exception(() => engine.RunTarget("A"));
+                var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: remove trailing space or line break after `=> `

**Pull request merged**